[New Tutorial] Solving Problems You Can't See: Using AWS X-Ray and CloudWatch for user-level observability in your serverless microservices applications

knizami opened the pull request:
# Content Pull Request Checklist

Please check to make sure your content submission meets the following requirements by adding an `x` between the brackets:

- [x ] The title of this pull request reflects the title of my content.
- [x ] I have reviewed the [Content Review Checklist](/content/blob/main/CONTENT_REVIEW_CHECKLIST.md).
- [x ] This pull request contains one piece of content (post or page).

**Important: if you are submitting a PR to update existing content, please ensure you [sync'ed your fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) first before you created your branch.**

Is this pull request linked to a Content Proposal or Idea Suggestion Issue? If so, please link to it below using the number or full link.

**By submitting this pull request, I confirm I own this content, have permission to use each image included in this content, and that it will be released under the [CC BY-SA 4.0 SA License](/LICENSE).**

knizami commented:
Quip version is here.  Reviewed by @bketelsen: 

https://quip-amazon.com/mEFvAjXX8MmX/How-to-group-measure-and-alarm-on-user-requests-in-a-serverless-microservices-architecture-using-X-Ray-and-CloudWatch-WORKING-DRAFT

knizami commented:
One question on the port.  I was highlighting parts of my code blocks in Quip but I don't see a way to do this in markdown.  Let me know your thoughts on this.

knizami commented:
Also, this may be more appropriate as a tutorial given its length, walkthrough guidance (reproducibility), and flow?

scribwell commented:
@bketelsen Would you please perform the technical review for this article? Also, we'll need to reformat this as a tutorial as it was incorrectly classified as a blog post. Please work with the author to ensure this article is using the correct format.

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1111 @@
+
+```
+
+import ***CartButton* **from "@/components/CartButton.vue";
````
remove all the extraneous `*` symbols here. the syntax highlighter will highlight these appropriately if they're annotated with the `javascript` tag in the codeblock.

knizami replied:
done.

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1111 @@
+
+This file provides the layout for the application.  You can see that the main product content for the application is being loaded from the router and rendered in **<router-view />**.  You can also see that the cart drawer (<cart-drawer />) **** remains constant, even if the path and route for the application changes.  When the page is first loaded, the shopping cart for the user is retrieved:
+
+```
````
and remove all the extra `*`'s here too!

knizami replied:
done.

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1111 @@
+```
+export async function getCart() {
+    return getHeaders(true).then(
+        headers => **API****.get("CartAPI", "/cart", {
````
remove extra `*`'s here too.

knizami replied:
done.

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
```diff
@@ -0,0 +1,1111 @@
+
+<script>
+export default {
+**  computed: {
```
remove extra `*` here too

knizami replied:
done.

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
```diff
@@ -0,0 +1,1111 @@
+  },
+  API: {
+    endpoints: [{
+**        name: "CartAPI",
```
remove these `*`

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1111 @@
+
+```
+@logger.inject_lambda_context(log_event=True)
+**@tracer.capture_lambda_handler
````
remove these *

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1111 @@
+```
+export async function getCart() {
+    return getHeaders(true).then(
+        headers => **API****.get("CartAPI", "/cart", {
````
remove these *

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
```diff
@@ -0,0 +1,1111 @@
+import os
+
+import boto3
+**from aws_xray_sdk.core import xray_recorder
```
remove these `*`

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1111 @@
+
+Update your **backend/shopping-cart-service/list_cart.py** file to the following:
+
+```
````
add `python` tag here. like ```python

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
```diff
@@ -0,0 +1,1111 @@
+    List items in shopping cart.
+    """
+ **subsegment = xray_recorder.begin_subsegment('annotations')
+**
```
remove all these `*`

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1111 @@
+
+Next, update the **backend/layers/shared.py** file to the following:
+
+```
````
add `python` annotation here

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
```diff
@@ -0,0 +1,1111 @@
+
+    return verified_claims.get("sub")
+
+**@tracer.capture_method
```
remove * here

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1111 @@
+Before you go fix this issue, take a closer look at the logs provided by AWS X-Ray.  One thing to note here is that the logs displayed in X-Ray don’t contain the exception as you would expect it to.  You will notice in the screenshot above that the log statements that are included in X-Ray include the X-Ray trace id.  AWS X-Ray is using [CloudWatch Logs Insights](https://docs.aws.amazon.com/Amazo ...
+
+```
+fields *@log*, *@timestamp*, *@message*
````
remove the *'s here

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1111 @@
+
+```
+fields *@log*, *@timestamp*, *@message*
+| filter *@message* like "05bf2f42-cd3b-4e4b-8876-f97b022525be" or @requestId = "c2eac3ec-d724-496c-9eb0-a112dc2c459d" or *@message* like "1-6470a77f-746c4b830dc79084524259da"
````
and here

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1111 @@
+```
+fields *@log*, *@timestamp*, *@message*
+| filter *@message* like "05bf2f42-cd3b-4e4b-8876-f97b022525be" or @requestId = "c2eac3ec-d724-496c-9eb0-a112dc2c459d" or *@message* like "1-6470a77f-746c4b830dc79084524259da"
+| sort *@timestamp*, *@message* desc
````
and here

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1111 @@
+to:
+
+```
+**    for index, product in enumerate(product_list):
````
remove all these `*`

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1111 @@
+from:
+
+
+```
````
add python tag here

knizami replied:
done.

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1111 @@
+to:
+
+
+```
````
add python tag here

knizami replied:
done.

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1111 @@
+
+
+```
+**    try:
````
remove **'s throughout

knizami replied:
done.

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
```diff
@@ -0,0 +1,1111 @@
+from boto3.dynamodb.conditions import Key
+
+from shared import get_cart_id, get_headers, get_user_sub, get_username, handle_decimal_type
+**from time import sleep
```
remove **'s throughout

knizami replied:
done.

bketelsen commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1111 @@
+
+Call this group **SlowRegisteredUsers** and save the group.  Now, slow down the **aws-serverless-shopping-cart-shop-ListCartFunction-<hash>** AWS Lambda function by updating the **backend/shopping-cart-service/list_cart.py** file:
+
+```
````
add python annotation

knizami replied:
done.

bketelsen reviewed (changes requested):
This is an excellent and thorough article on X-ray and Cloudwatch. Left several comments, mostly related to code formatting.  Please remove all the `*`'s in the code blocks and annotate all the code blocks with the language used.

bketelsen commented:
Needs to be moved to tutorials folder, use correct template.

knizami commented on posts/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1111 @@
+This script creates the Vue application and loads the Vue components that make up the application.  You can see that each component is loaded from a separate file in the **frontend/src/components** directory.  You will focus on the shopping cart (CartDrawer) and product listing (Product) components.
+
+
+```
````
done.

knizami commented:
I've moved this to the tutorial section and cleaned up the content.  I'm going to do another dry run to make sure everything flows well and provide some additional minor edits.

bketelsen commented:
@knizami please remove the `.idea` folder that was committed, these are IDE specific files for your editor.

scribwell commented:
@damonk10 I've assigned this tutorial to 8112 for editorial review.

DalM2023 commented on tutorials/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/images/x-ray-service-map-node-view.png:
I can't see this image. It's unavailable.

DalM2023 commented on tutorials/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/images/x-ray-service-map.png:
Screenshot of Service map with diagram of many possible scenarios starting from client icon to Application stage. From there, possible scenarios lead to Lambda content and Lambda Function and Dynamo.

DalM2023 commented on tutorials/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/images/x-ray-traces-list.png:
Image of Traces table showing OK status.

DalM2023 commented on tutorials/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/images/x-ray-traces-map.png:
Image of Trace Map with Legend and options box, highlighting Lambda Function.

DalM2023 commented on tutorials/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/images/x-ray-traces-view.png:
Image of Traces table showing OK status.

DalM2023 commented on tutorials/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/images/x-ray-view-filtered-traces-select-segment-with-error.png:
Image of Segments Timeline highlighting Lamda link with Error warning.

DalM2023 commented on tutorials/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/images/x-ray-view-filtered-traces.png:
Image showing segment details of shopping cart, highlighting Fault.

DalM2023 commented on tutorials/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
```diff
@@ -0,0 +1,1247 @@
+---
+title: Solving Problems You Can't See - Using AWS X-Ray and CloudWatch for user-level observability in your serverless microservices applications
```
```suggestion
title: Solve Problems You Can't See- How AWS X-Ray and CloudWatch Help You Save Stranded Users
```

knizami replied:
Hi, I think its important that we include "observability" and "serverless microservices applications" as these statements provide specificity for the tutorial.   We could shorten it to:
Solving Problems You Can't See- How AWS X-Ray and CloudWatch for can provide observability in your serverless microservices applications
 
or something similar

DalM2023 replied:
Got it. So, please rewrite the descriptor. The title and the descriptor don't need to repeat each other. 

Keep the title in AP Style: "Solving Problems You Can’t See- How AWS X-Ray and CloudWatch Provide Observability in Serverless Microservices Applications".

Then, rewrite the descriptor without the words "observability" and "serverless microservices applications". It should also be written at 60 characters. 

Something like: 
EX: Avoid blind spots by tracking users on apps built on AWS.

DalM2023 commented on tutorials/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
```diff
@@ -0,0 +1,1247 @@
+---
+title: Solving Problems You Can't See - Using AWS X-Ray and CloudWatch for user-level observability in your serverless microservices applications
+description: Learn how to use AWS X-Ray to trace and observe your user interactions in a serverless microservices application built on AWS.
```
```suggestion
description: Observe users in serverless microservices applications built on AWS.
```

DalM2023 commented on tutorials/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
```diff
@@ -0,0 +1,1247 @@
+---
```
Hello! This is an excellent blog! Great job with the intro. You set up the scene nicely and it invites the reader. The structure is solid too. You have a friendly voice and conversational tone in the piece that plays well with the aphorisms. 

For my edits I added contractions here and there to make it more conversational. I tightened some sentences to eliminate overuse of "that", so it reads more fluidly. 

Since the links are embedded into the copy I removed the brackets ONLY when the links are part of a complete sentence. It makes the blog easier to read digitally. I also capitalized titles and proper nouns. 

Overall, well done!

knizami replied:
Thank you for all your edits.  The only concern I have is with making the title too general because then its harder for the reader to understand what they will learn.  Also I believe there were 2 images that you found that you said were unavailable?  I am not sure what the issue is there, these are showing up on my version.

DalM2023 replied:
See my notes below on the title and descriptor. 

I still can't see all the images. No idea why. But I trust your end's technology more than mine.

scribwell commented:
@knizami Please review our our editor's suggestions/revisions and let us know when you've completed all revisions so we can send this through final review and get it published!

knizami commented on tutorials/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/images/aws-x-ray-cloudwatch-group-count-create-alarm-2.png:
Hi not sure about this comment?

knizami commented:
> 

editorial changes reviewed.

DalM2023 commented on tutorials/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1172 @@
+| sort @timestamp, @message desc
+```
+
+You can see that the query is looking for either the API Gateway request id, AWS Lambda Function request id, or the AWS X-Ray trace id in the message.   When you log messages with the **aws_lambda_powertools** logger and you have also instrumented your function to use AWS X-Ray, it will automatically enrich the log messages with the AWS X-Ray trace id.  However, the exception that was thrown was  ...
````
```suggestion
You can see that the query is looking for either the API Gateway request id, AWS Lambda Function request id, or the AWS X-Ray trace ID in the message.   When you log messages with the **aws_lambda_powertools** logger and you have also instrumented your function to use AWS X-Ray, it will automatically enrich the log messages with the AWS X-Ray trace ID.  However, the exception that was thrown was not logged using the logger and therefore didn’t include the X-Ray trace id.  You can confirm this by proceeding to the logs for the **aws-serverless-shopping-cart-shop-ListCartFunction-\<hash>** AWS Lambda function and viewing the exception message in the logs:  
```

DalM2023 commented on tutorials/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
````diff
@@ -0,0 +1,1172 @@
+    xray_recorder.end_subsegment()
+```
+
+The exception was caused by incorrectly using the `remove()` python list function on the individual product dictionary.  Next, you will fix this error and also improve the exception handling for the function by wrapping the execution in a try / except block:
````
```suggestion
The exception was caused by incorrectly using the `remove()` python list function on the individual product dictionary.  Next, you will fix this error and also improve the exception handling for the function by wrapping the execution in a try / except block:
```
```suggestion
The exception was caused by incorrectly using the `remove()` Python list function on the individual product dictionary.  Next, you will fix this error and also improve the exception handling for the function by wrapping the execution in a try / except block:
```

DalM2023 reviewed (commented):
Second review done. Ready for review.

scribwell commented:
@knizami We have a few more author revisions/edits for you to review before we can publish. Would you please review?

knizami commented on tutorials/solving-problems-you-cant-see-using-aws-x-ray-and-cloudwatch-for-user-level-observability-in-your-serverless-microservices-applications/index.md:
```diff
@@ -0,0 +1,1247 @@
+| 📢 Feedback            | <a href="https://pulse.buildon.aws/survey/DEM0H5VW" target="_blank">Any feedback, issues, or just a</a> 👍 / 👎 ?       |
+| ⏰ Last Updated     | 2023-09-02                                                                                                              |
+
+| ToC |
```
Hi, not sure @scribwell ?

knizami commented:
Thanks @DalM2023 for the excellent review / update.  I incorporated your changes for the title / descriptor.  @scribwell can we confirm that images that @DalM2023 is having problems viewing are not a problem in general?  I am able to see all the images load.

scribwell commented:
@damonk10 Please perform the final review and ensure there are no issues with the images before publishing.
